sql/analyzer/optimization_rules.go

```diff
@@ -149,21 +149,40 @@ func moveJoinConditionsToFilter(ctx *sql.Context, a *Analyzer, n sql.Node, scope
 		return node, transform.SameTree, nil
 	}
 
-	return transform.NodeWithCtx(node, func(transform.Context) bool { return true }, func(c transform.Context) (sql.Node, transform.TreeIdentity, error) {
-		if c.Node != topJoin {
-			return c.Node, transform.SameTree, nil
+	resultNode, resultIdentity, err := transform.Node(node, func(n sql.Node) (sql.Node, transform.TreeIdentity, error) {
+		children := n.Children()
+		if children == nil || len(children) == 0 {
+			return n, transform.SameTree, nil
+		}
+		isParentOfTopJoin := false
+		for _, child := range children {
+			if child == topJoin {
+				isParentOfTopJoin = true
+				break
+			}
 		}
-		switch n := c.Parent.(type) {
+		if !isParentOfTopJoin {
+			return n, transform.SameTree, nil
+		}
+
+		switch imp := n.(type) {
 		case *plan.Filter:
-			return plan.NewFilter(
-				expression.JoinAnd(append([]sql.Expression{n.Expression}, nonJoinFilters...)...),
-				n.Child), transform.NewTree, nil
+			newExpression := expression.JoinAnd(append([]sql.Expression{imp.Expression}, nonJoinFilters...)...)
```

Comment:
this might be a bit shorter
```suggestion
			newExpression := expression.JoinAnd(append(nonJoinFilters, n.Expression)...)
```

Reply:
Shorter, and we're also not expanding `nonJoinFilters` unnecessarily. Changing.